lib/snippets/deploy-to-gke

```diff
@@ -0,0 +1,142 @@
+    @gcloud_deployment = gcloud_deployment
+    @gcloud_project = gcloud_project
+    @current_sha = current_sha
+    enforce_required_parmas
+    @id = current_sha + "-#{SecureRandom.hex(4)}"
+    @template_path = './' + template_folder || "config/deploy/#{environment}"
+  end
+
+  def run
+    authorize_gcloud
+    fetch_clusters.each do |cluster|
+      set_kubectl_cluster(cluster)
+      apply_all_templates
+    end
+  rescue FatalDeploymentError => error
+    print_error(error.message)
+    exit 1
+  end
+
+  def template_variables
+    {
+      'current_sha' => @current_sha,
+      'deployment_id' => @id,
+    }
+  end
+
+  private
+
+  def enforce_required_parmas
+    [@namespace, @gcloud_project, @gcloud_deployment, @current_sha].each do |required_param|
+      raise ArgumentError, "#{required_param} is required" unless required_param && !required_param.empty?
+    end
+  end
+
+  def apply_all_templates
+    unless File.directory?(@template_path)
+      raise FatalDeploymentError, "#{@template_path} doesn't exist"
+    end
+    found = 0
+    Dir.foreach(@template_path) do |file|
+      file_path = "#{@template_path}/#{file}"
+      if File.extname(file) == '.yml'
+        found += 1
+        run_command('kubectl', 'apply', '-f', file_path, "--namespace=#{@namespace}")
+      elsif File.extname(file) == '.erb'
+        found += 1
+        render_and_apply_template(file_path)
+      end
+    end
+    raise FatalDeploymentError, "No templates found in #{@template_path}" if found.zero?
+  end
+
+  def render_and_apply_template(file_path)
+    erb_template = ERB.new(File.read(file_path))
+    erb_binding = binding
+    template_variables.each do |var_name, value|
+      erb_binding.local_variable_set(var_name, value)
+    end
+    content = erb_template.result(erb_binding)
+
+    f = Tempfile.new(['kube_template', '.yml'])
+    f.write(content)
+    f.close
+    run_command('kubectl', 'apply', '-f', f.path, "--namespace=#{@namespace}")
+  ensure
+    f.unlink if f
+  end
+
+  def authorize_gcloud
+    status = run_command('gcloud', '-q', 'auth', 'activate-service-account', '--key-file', "#{ENV['GCLOUD_CREDENTIALS_DIR']}/#{@gcloud_project}.json")
+    raise FatalDeploymentError, "Failed to set gcloud project #{@gcloud_project}" unless status
+  end
+
+  def set_kubectl_cluster(cluster)
+    cluster_name = cluster[0]
+    cluster_zone = cluster[1]
+    status = run_command('gcloud', '-q', 'container', 'clusters', 'get-credentials', cluster_name, '--zone', cluster_zone)
+    raise FatalDeploymentError, "Failed to set cluster #{cluster_name}/#{cluster_zone}" unless status
+  end
+
+  def fetch_clusters
+    result = query_deployments
+    result['resources'].each_with_object([]) do |resource, data|
+      next unless resource['type'] == 'container.v1.cluster'
+      properties = YAML.load(resource['finalProperties'])
+      data << [resource['name'], properties['zone']]
+    end
+  end
+
+  def run_command(*args)
+    puts args.join(" ")
```

Comment:
I'd use `Shellwords.join` so that if we ever have arguments with spaces or quotes, this output won't be misleading as of what command was ran exactly.